lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,181 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @stderr, @ps_process = Open3.popen3(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code, timeout_ms = 300 * 1000)
+        output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        output_ready_event = self.class.create_event(output_ready_event_name)
+
+        code = make_ps_code(powershell_code, output_ready_event_name, timeout_ms)
+
+        out, err = exec_read_result(code, output_ready_event)
+
+        { :stdout => out, :stderr => err }
```

Comment:
I think that we should reshape the object here to make a bit more sense, like this:

```ruby
out = JSON.parse(out)

{
  :exitcode => out['exitcode'],
  :errormessage => out['errormessage'],
  :stdout => out['output'],
  :stderr => err
}
```

Then the downstream consumer doesn't need to do any JSON parsing, and the members are really indicative of what they are.